src/clientsettingsdlgbase.ui

```diff
@@ -358,6 +358,31 @@
              </property>
             </spacer>
            </item>
+           <item>
```

Comment:
Hm.  This doesn't look right.  I prefer JACK's approach of listing all the devices as "API: device driver name".  Then the existing panel layout can remain and the command line argument isn't needed, making it easy to use.

Reply:
This doesn't account for the shared vs exclusive of WASAPI mode though (which is what the code you are commenting is about)?

I also dispute that multiplying the number of device entries by (up to) 3 will make things easier. Having n^2 entries due to the Cartesian product of input and output devices is already bad enough.

Reply:
Why can't they be returned as a single driver's list of inputs and outputs, anyway, rather than having to be returned each as a separate driver as they are now?

Reply:
I don't understand what you mean.

Reply:
OK, go about it more list ASIO - but, rather than separating out each hardware device, have the PortAudio API be the device.  This way you could choose the API without having to have `--api`.

Then you have Input 1, Input 2, Output 1 and Output 2 boxes.  For the chosen API, you'd return all the inputs to populate the input boxes and all the outputs to populate the output boxes.

Benefits:
* No cartesian product.
* Similar to the existing user experience.
* No command line option required.

Reply:
Oh, I see. I guess that works as long as the devices themselves don't have multiple channels, I think this is usually the case for non-ASIO devices. I'm wouldn't be too happy about special-casing the code like that. Also it still doesn't solve the shared vs exclusive WASAPI mode thing.

> * Similar to the existing user experience.

The Cartesian product thing is the existing user experience on macOS, by the way.

Reply:
I'm explicitly referring to ASIO here as the primary target for this change is Windows ASIO4ALL users.